File: LeetCode/23.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None
        
        while len(lists) > 1:
            merge_result = []

            for i in range(0, len(lists), 2):
                l1 = lists[i]
                if i + 1 < len(lists):
                    l2 = lists[i+1]
                else:
                    l2 = None
                merge_result.append(self.mergeLists(l1, l2))

            lists = merge_result
        
        return lists[0]

    def mergeLists(self, l1, l2):
        dummy = ListNode()
        curr = dummy

        while l1 or l2:
            if l1 and l2:
                if l2.val < l1.val:
                    curr.next = ListNode(l2.val)
                    l2 = l2.next
                else:
                    curr.next = ListNode(l1.val)
                    l1 = l1.next

            elif l1 and not l2:
                curr.next = ListNode(l1.val)
                l1 = l1.next
                
            else:
                curr.next = ListNode(l2.val)
                l2 = l2.next
                
            curr = curr.next

        return dummy.next
                

File: leetcode-spy/23.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *&merge(ListNode *&c1, ListNode *&c2)
    {
        ListNode *dummy = new ListNode();
        ListNode *curr = dummy;
        while (c1 && c2)
        {
            if (c1->val <= c2->val)
            {
                curr->next = new ListNode(c1->val);
                c1 = c1->next;
            }
            else
            {
                curr->next = new ListNode(c2->val);
                c2 = c2->next;
            }
            curr = curr->next;
        }
        if (c1)
        {
            curr->next = c1;
        }
        if (c2)
        {
            curr->next = c2;
        }
        return dummy->next;
    }

    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        if (lists.empty())
            return nullptr;

        while (lists.size() > 1)
        {
            vector<ListNode *> tmp;
            for (int i = 0; i < lists.size(); i += 2)
            {
                ListNode *l1 = lists[i];
                ListNode *l2 = i + 1 < lists.size() ? lists[i + 1] : nullptr;
                tmp.push_back(merge(l1, l2));
            }
            lists = move(tmp);
        }
        return lists[0];
    }
};
